src/serial.cpp

```diff
@@ -0,0 +1,332 @@
+    return B300;
+  case 600:
+    return B600;
+  case 1200:
+    return B1200;
+  case 1800:
+    return B1800;
+  case 2400:
+    return B2400;
+  case 4800:
+    return B4800;
+  case 9600:
+    return B9600;
+  case 19200:
+    return B19200;
+  case 38400:
+    return B38400;
+  case 57600:
+    return B57600;
+  case 115200:
+    return B115200;
+  case 230400:
+    return B230400;
+#ifdef B460800
+  case 460800:
+    return B460800;
+#endif
+#ifdef B921600
+  case 921600:
+    return B921600;
+#endif
+  default:
+    return B9600; // fallback
+  }
+}
+
+inline void invokeError(int code) {
+  if (errorCallback) {
+    errorCallback(code);
+  }
+}
+
+} // namespace
+
+// -----------------------------------------------------------------------------
+// Public API implementation
+// -----------------------------------------------------------------------------
+
+intptr_t serialOpen(void *port, int baudrate, int dataBits, int parity,
+                    int stopBits) {
+  if (!port) {
+    invokeError(std::to_underlying(StatusCodes::INVALID_HANDLE_ERROR));
+    return 0;
+  }
+
+  const char *portName = static_cast<const char *>(port);
+  int fd = open(portName, O_RDWR | O_NOCTTY | O_SYNC);
+  if (fd < 0) {
+    invokeError(std::to_underlying(StatusCodes::INVALID_HANDLE_ERROR));
+    return 0;
+  }
+
+  SerialPortHandle *handle = new SerialPortHandle{fd, {}};
+
+  termios tty{};
+  if (tcgetattr(fd, &tty) != 0) {
+    invokeError(std::to_underlying(StatusCodes::GET_STATE_ERROR));
+    close(fd);
+    delete handle;
+    return 0;
+  }
+  handle->original = tty; // save original
+
+  // Basic flags: local connection, enable receiver
+  tty.c_cflag |= (CLOCAL | CREAD);
+
+  // Baudrate
+  const speed_t speed = to_speed_t(baudrate);
+  cfsetispeed(&tty, speed);
+  cfsetospeed(&tty, speed);
+
+  // Data bits
+  tty.c_cflag &= ~CSIZE;
+  switch (dataBits) {
+  case 5:
+    tty.c_cflag |= CS5;
+    break;
+  case 6:
+    tty.c_cflag |= CS6;
+    break;
+  case 7:
+    tty.c_cflag |= CS7;
+    break;
+  default:
+    tty.c_cflag |= CS8;
+    break;
+  }
+
+  // Parity
+  if (parity == 0) {
+    tty.c_cflag &= ~PARENB;
+  } else {
+    tty.c_cflag |= PARENB;
+    if (parity == 1) {
+      tty.c_cflag &= ~PARODD; // even
+    } else {
+      tty.c_cflag |= PARODD; // odd
+    }
+  }
+
+  // Stop bits
+  if (stopBits == 2) {
+    tty.c_cflag |= CSTOPB;
+  } else {
+    tty.c_cflag &= ~CSTOPB;
+  }
+
+  // Raw mode (no echo/processing)
+  tty.c_iflag = 0;
+  tty.c_oflag = 0;
+  tty.c_lflag = 0;
+
+  tty.c_cc[VMIN] = 0;   // non-blocking by default
+  tty.c_cc[VTIME] = 10; // 1s read timeout
+
+  if (tcsetattr(fd, TCSANOW, &tty) != 0) {
+    invokeError(std::to_underlying(StatusCodes::SET_STATE_ERROR));
+    close(fd);
+    delete handle;
+    return 0;
+  }
+
+  return reinterpret_cast<intptr_t>(handle);
+}
+
+void serialClose(int64_t handlePtr) {
+  auto *handle = reinterpret_cast<SerialPortHandle *>(handlePtr);
+  if (!handle)
+    return;
```

Comment:
Please don't use IF-Statements without `{}`

```suggestion
  if (!handle) {
    return;
  }
```